src/android/com/microsoftopentechnologies/intellij/wizards/createvm/EndpointStep.java

```diff
@@ -0,0 +1,396 @@
+        endpointsTable.getColumnModel().getColumn(4).setPreferredWidth(60);
+        endpointsTable.getColumnModel().getColumn(5).setPreferredWidth(205);
+
+        endpointsTable.setTableHeader(null);
+
+
+        endpointsTable.addMouseListener(new MouseAdapter() {
+            @Override
+            public void mouseClicked(MouseEvent evt) {
+                int row = endpointsTable.rowAtPoint(evt.getPoint());
+                int col = endpointsTable.columnAtPoint(evt.getPoint());
+                if (col == 5) {
+                    //Determine if click was on the "X" label
+                    if(endpointsTable.getWidth() - 30 < evt.getX()) {
+                        endpointTableModel.getData().remove(row);
+                        endpointTableModel.fireTableDataChanged();
+                    }
+                }
+            }
+        });
+
+        endpointTableModel.addTableModelListener(new TableModelListener() {
+            @Override
+            public void tableChanged(TableModelEvent tableModelEvent) {
+
+                boolean hasErrors = false;
+                for (int i = 0; i < endpointTableModel.getRowCount(); i++) {
+                    String errorFromRow = getErrorFromRow(i, endpointTableModel.getData());
+                    if(errorFromRow.length() > 0)
+                        hasErrors = true;
+                }
+
+                model.getCurrentNavigationState().FINISH.setEnabled(!hasErrors && endpointTableModel.getRowCount() > 0);
+            }
+        });
+    }
+
+    @Override
+    public JComponent prepare(WizardNavigationState wizardNavigationState) {
+        rootPanel.revalidate();
+        return rootPanel;
+    }
+
+    @Override
+    public boolean onFinish() {
+
+        final EndpointTableModel tableModel = (EndpointTableModel) endpointsTable.getModel();
+        ProgressManager.getInstance().run(new Task.Backgroundable(project, "Creating virtual machine...", false) {
+
+            @Override
+            public void run(ProgressIndicator progressIndicator) {
+                progressIndicator.setIndeterminate(true);
+
+                try {
+
+
+                    final VirtualMachine virtualMachine = new VirtualMachine(
+                            model.getName(),
+                            model.getCloudService().getName(),
+                            model.getCloudService().getProductionDeployment(),
+                            model.getAvailabilitySet(),
+                            model.getSize().getName(),
+                            "",
+                            model.getSubscription().getId().toString()
+                    );
+
+                    virtualMachine.getEndpoints().addAll(tableModel.getData());
+
+                    AzureSDKManagerImpl.getManager().createVirtualMachine(virtualMachine, model.getVirtualMachineImage(), model.getStorageAccount(), model.getUserName(), new String(model.getPassword()));
```

Comment:
Should we be creating the virtual machine here? I'd think the right place to do this would be the action listener for the menu item, i.e. in the `CreateVMAction` class, no?